contrib/mesos/pkg/scheduler/podtask/procurement.go

```diff
@@ -0,0 +1,145 @@
+)
+
+// NewDefaultProcurement returns the default procurement strategy that combines validation,
+// node, pod-resource, and ports procurement. c and m are resource quantities written into
+// k8s api.Pod.Spec's that don't declare resources (all containers in k8s-mesos require cpu
+// and memory limits).
+func NewDefaultProcurement(c mresource.CPUShares, m mresource.MegaBytes) Procurement {
+	requireSome := &RequireSomePodResources{
+		defaultContainerCPULimit: c,
+		defaultContainerMemLimit: m,
+	}
+	return AllOrNothingProcurement([]Procurement{
+		ValidateProcurement,
+		NodeProcurement,
+		requireSome.Procure,
+		PodResourcesProcurement,
+		PortsProcurement,
+	}).Procure
+}
+
+// Procurement funcs allocate resources for a task from an offer.
+// Both the task and/or offer may be modified.
+type Procurement func(*T, *mesos.Offer) error
+
+type AllOrNothingProcurement []Procurement
+
+func (a AllOrNothingProcurement) Procure(t *T, details *mesos.Offer) error {
+	for _, p := range a {
+		if err := p(t, details); err != nil {
+			t.Reset()
+			return err
+		}
+	}
+	return nil
+}
+
+// ValidateProcurement checks that the offered resources are kosher, and if not panics.
+// If things check out ok, t.Spec is cleared and nil is returned.
+func ValidateProcurement(t *T, details *mesos.Offer) error {
+	if details == nil {
+		//programming error
+		panic("offer details are nil")
+	}
+	t.Spec = Spec{}
+	return nil
+}
+
+// NodeProcurement updates t.Spec in preparation for the task to be launched on the
+// slave associated with the offer.
+func NodeProcurement(t *T, details *mesos.Offer) error {
+	t.Spec.SlaveID = details.GetSlaveId().GetValue()
+	t.Spec.AssignedSlave = details.GetHostname()
+
+	// hostname needs of the executor needs to match that of the offer, otherwise
+	// the kubelet node status checker/updater is very unhappy
+	const HOSTNAME_OVERRIDE_FLAG = "--hostname-override="
+	hostname := details.GetHostname() // required field, non-empty
+	hostnameOverride := HOSTNAME_OVERRIDE_FLAG + hostname
+
+	argv := t.executor.Command.Arguments
+	overwrite := false
+	for i, arg := range argv {
+		if strings.HasPrefix(arg, HOSTNAME_OVERRIDE_FLAG) {
+			overwrite = true
+			argv[i] = hostnameOverride
+			break
+		}
+	}
+	if !overwrite {
+		t.executor.Command.Arguments = append(argv, hostnameOverride)
+	}
+	return nil
+}
+
+type RequireSomePodResources struct {
+	defaultContainerCPULimit mresource.CPUShares
+	defaultContainerMemLimit mresource.MegaBytes
+}
+
+func (r *RequireSomePodResources) Procure(t *T, details *mesos.Offer) error {
+	// write resource limits into the pod spec which is transferred to the executor. From here
+	// on we can expect that the pod spec of a task has proper limits for CPU and memory.
+	// TODO(sttts): For a later separation of the kubelet and the executor also patch the pod on the apiserver
+	if unlimitedCPU := mresource.LimitPodCPU(&t.Pod, r.defaultContainerCPULimit); unlimitedCPU {
```

Comment:
this is also a change outside of t.spec. Not sure I like it here.

We could create a deep copy of t.Pod in t.Spec to clean this up.